pkg/cmd/cli/cmd/observe/observe.go

```diff
@@ -0,0 +1,957 @@
+	%[1]s observe services
+
+	# observe changes to services, including the clusterIP and invoke a script for each
+	%[1]s observe services -a '{ .spec.clusterIP }' --each register_dns.sh
+`
+)
+
+// NewCmdObserve creates the observe command.
+func NewCmdObserve(fullName string, f *clientcmd.Factory, out, errOut io.Writer) *cobra.Command {
+	options := &ObserveOptions{
+		baseCommandName: fullName,
+		retryCount:      2,
+		templateType:    "jsonpath",
+		maximumErrors:   20,
+		listenAddr:      ":11251",
+	}
+
+	cmd := &cobra.Command{
+		Use:     "observe",
+		Short:   "Observe changes to resources and invoke integrations",
+		Long:    fmt.Sprintf(observeLong, fullName),
+		Example: fmt.Sprintf(observeExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := options.Complete(f, args, out, errOut); err != nil {
+				cmdutil.CheckErr(err)
+			}
+
+			if err := options.Validate(args); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.Run(); err != nil {
+				cmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	// flags controlling what to select
+	cmd.Flags().BoolVar(&options.allNamespaces, "all-namespaces", false, "If present, list the requested object(s) across all projects. Project in current context is ignored.")
+
+	// the commands to execute on changes
+	cmd.Flags().VarP(&options.eachCommand, "each", "e", "A command to run for each change. Specify mulitple times to add arguments.")
+
+	// to perform deletion synchronization
+	cmd.Flags().VarP(&options.deleteCommand, "delete", "d", "A command to run when resources are deleted. Specify mulitple times to add arguments.")
+	cmd.Flags().Var(&options.nameSyncCommand, "names", "A command that will list all of the currently known names, optional. Specify multiple times to add arguments. Use to get notifications when objects are deleted.")
+
+	// add additional arguments / info to the server
+	cmd.Flags().StringVar(&options.templateType, "use", options.templateType, "Format for template arguments. jsonpath and gotemplate are supported.")
+	cmd.Flags().BoolVar(&options.strictTemplates, "strict-templates", false, "If true return an error on any field or map key that is not missing in a template.")
+	cmd.Flags().VarP(&options.templates, "argument", "a", "Template for the arguments to be passed to each command in the format defined by --use.")
+	cmd.Flags().StringVar(&options.typeEnvVar, "type-env-var", "", "The name of an env var to set with the type of event received ('Sync', 'Updated', 'Deleted', 'Added') as the first argument to the --each command.")
+	cmd.Flags().StringVar(&options.objectEnvVar, "object-env-var", "", "The name of an env var to serialize the object to when calling --each, optional.")
+
+	// control retries of individual commands
+	cmd.Flags().IntVar(&options.maximumErrors, "maximum-errors", options.maximumErrors, "Exit after this many errors have been detected with. May be set to -1 for no maximum.")
```

Comment:
The point of this is to get you to fix your bugs.  The worst outcome to me
is you have a lot of uncovered edge cases and you never discover it because
you don't look at your logs.  The best outcome is by default you get a nice
error "hey, you hit too many errors, maybe you should think about fixing
your stuff".  I am worried about hitting the limit on sync and not getting
through the whole list (so maybe we can defer the exit until after the
first sync is complete). OTOH if you have > 50 items and you fail the first
twenty I don't see much point to continuing.

I think of this as gamifying error handling - you are a new user, we will
gently nudge you to do things that handle errors, without trying to
surprise you.

On Tue, Aug 30, 2016 at 3:52 PM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/observe/observe.go
> <https://github.com/openshift/origin/pull/4196#discussion_r76866832>:
>
> > +	// the commands to execute on changes
> > +	cmd.Flags().VarP(&options.eachCommand, "each", "e", "A command to run for each change. Specify mulitple times to add arguments.")
> > +
> > +	// to perform deletion synchronization
> > +	cmd.Flags().VarP(&options.deleteCommand, "delete", "d", "A command to run when resources are deleted. Specify mulitple times to add arguments.")
> > +	cmd.Flags().Var(&options.nameSyncCommand, "names", "A command that will list all of the currently known names, optional. Specify multiple times to add arguments. Use to get notifications when objects are deleted.")
> > +
> > +	// add additional arguments / info to the server
> > +	cmd.Flags().StringVar(&options.templateType, "use", options.templateType, "Format for template arguments. jsonpath and gotemplate are supported.")
> > +	cmd.Flags().BoolVar(&options.strictTemplates, "strict-templates", false, "If true return an error on any field or map key that is not missing in a template.")
> > +	cmd.Flags().VarP(&options.templates, "argument", "a", "Template for the arguments to be passed to each command in the format defined by --use.")
> > +	cmd.Flags().StringVar(&options.typeEnvVar, "type-env-var", "", "The name of an env var to set with the type of event received ('Sync', 'Updated', 'Deleted', 'Added') as the first argument to the --each command.")
> > +	cmd.Flags().StringVar(&options.objectEnvVar, "object-env-var", "", "The name of an env var to serialize the object to when calling --each, optional.")
> > +
> > +	// control retries of individual commands
> > +	cmd.Flags().IntVar(&options.maximumErrors, "maximum-errors", options.maximumErrors, "Exit after this many errors have been detected with. May be set to -1 for no maximum.")
>
> I don't think so. I think you're in a pod, and if you hit enough errors
> you restart. There is no sliding window that will work for all users. We
> could do % for sync but it wouldn't have any meaning after sync. In
> practice, your scripts should never error. If you're erroring, you want to
> restart.
>
> 10 errors out of 20 attempts is a lot different than 10 errors out of
> 20,000 attempts. Having one value is silly.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/4196/files/d21d6896f2faeac799f81a73286a7821db37b0d4#r76866832>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p_sCJ3E5qT64tInJ1Fmp0aecYoRuks5qlIoAgaJpZM4Fsaph>
> .
>